### PR TITLE
More than 2 hands!? WHAAAAAAT

### PR DESCRIPTION
**Pull request by RemieRichards**
## Changes:
- Mobs can now have more than 2 hands
- Minor rewrite of the "Inventory API" in terms of hands, this includes re-dos of procs, re-naming of procs, and refocusing hands around "held indexes"
- In terms of things like twohanded weapons/fake twohands like guns, arms act as "pairs", eg: the "upper left hand" is paired with the "upper right hand", this means if you're holding a gun with recoil in your upper left hand, the upper right hand is the one that's checked for emptiness
- for balance reasons you can still only wield 1 double wield weapon (I may remove this, since it'd not -too- much better)
- All mobs with inhands now draw them on 1 layer, like how some of the random `/carbon` and simples did, and how I tried to change it to almost 3 years ago when giacom told me no, there's literally no reason to have separate hand layers, all visual issues are the same, all visual effect is the same, it's just for unification. (yes it means you're potentially drawing double the images each time, but this is REALLY not costly at all as you'd have to be holding that many items too.)
- it "works" with dismemberment, it's just that all extra limbs are drawn where the normal ones would be, meaning you can't immediately tell how many hands somebody has, but attacking the "left arm" will keep attacking until they have no left arms, so functionally it all "works", visually: I need a spriter, probably.
- Being handcuffed handcuffs ALL your pairs of limbs, because video game logic.
## Todo:
- [x] Get a test merge
- [x] Fix crippling CPU usage from `get_held_index_of_item()` - Turns out, it _WAS_ actually lava's fault! @MrStonedOne has a PR up to fix that, so let's mark it fixed here too.
- [x] Fix Lizard/Species Accessories sprites - I have fixed (and future-proofed) them.
- [x] Fix Stacks (metal etc.) disappearing - Stacks could apparently `merge()` with themselves in some conditions, resulting in them consuming themselves (although it depended on the amount in the stack, which is confusing and really really !FUN!)
- [ ] Maybe get a 2nd test merge

:cl: 
add: Added the functionality for mobs to have multiple hands, this might have broken some things, so please report any weirdness
/:cl:

This is the first time I've seen all the changes laid out in front of me, so there are probably issues, but I should be able to sort them. I have tested this quite heavily, fixing any issues I've found but 2 servers full of players are more likely to find any rarer quirks. - <B>Edit:</B> looking at it, it all looks pretty good!

<del>Yes I know about the merge conflicts, it's probably somebody trying to use ```slot_[l/r]_hand``` which is no longer a thing OR one of the old proc names.</del>

## Screenshot:

old_screenshot
<B>Yes, the hands are backwards in the image. No, that's not an issue in the PR anymore</B>

### CODE REVIEW

File: code/__DEFINES/clothing.dm [hunk 1]
```diff
@@ -17,25 +17,25 @@
...
-#define slot_shoes			12
-#define slot_wear_suit		13
-#define slot_w_uniform		14
-#define slot_l_store		15
-#define slot_r_store		16
-#define slot_s_store		17
-#define slot_in_backpack	18
-#define slot_legcuffed		19
-#define slot_generic_dextrous_storage	20
+#define slot_hands			4 //wherever you provide a slot for hands you provide slot_hands AND a hand index (1 to held_items.len)
```
Comment (RemieRichards) [inline code/__DEFINES/clothing.dm]: this comment is wrong now, but hey, comments aren't top priority.

**Comment by yackemflam** [comment]
Fly people having 4 arms and 4 legs but are brittle as hell when?

**Comment by Jalleo** [comment]
Cue the need for four arm species!

**Comment by Jazaen** [comment]
No non-human roundstart shall have any advantage.
Also, does stripping/pickpocketing work on them?

File: code/_onclick/hud/swarmer.dm [hunk 2]
```diff
@@ -68,11 +68,11 @@
 	var/obj/screen/using
 
 	using = new /obj/screen/swarmer/FabricateTrap()
-	using.screen_loc = ui_rhand
+	using.screen_loc = ui_hand_position(2)
```
Comment (RemieRichards) [inline code/_onclick/hud/swarmer.dm]: the true solution to things like this would be to give the huds that cheated (read: used these ui positions incorrectly) is to give them their own ui positions, but for now, it's fine.

**Comment by RemieRichards** [comment]
@Jazaen Funnily enough, Stripping was something I didn't test, but It's a very simple change to the stripping panels, so it should work, but I think I may have missed 1 part of it, easy fix though.

**Comment by Nienhaus** [comment]
I already saw this yesterday B)

**Comment by RemieRichards** [comment]
Yep, people who donate to my patreon get sneak previews~

File: code/game/gamemodes/clock_cult/clock_unsorted.dm [hunk 3]
```diff
@@ -61,7 +61,7 @@
 	return ..()
 
 /datum/action/innate/function_call/Activate()
-	if(owner.l_hand && owner.r_hand)
+	if(!owner.get_empty_held_indexes())
 		usr << "<span class='warning'>You need an empty to hand to call forth your spear!</span>"
```
Comment (RemieRichards) [inline code/game/gamemodes/clock_cult/clock_unsorted.dm]: there's a typo here! it's not my typo though...

File: code/game/gamemodes/miniantags/abduction/machinery/console.dm [hunk 4]
```diff
@@ -173,7 +173,7 @@
 	entry.name = target.name
 	entry.icon = target.icon
 	entry.icon_state = target.icon_state
-	entry.overlays = target.get_overlays_copy(list(L_HAND_LAYER,R_HAND_LAYER))
+	entry.overlays = target.get_overlays_copy(list(HANDS_LAYER,HANDS_LAYER))
```
Comment (RemieRichards) [inline code/game/gamemodes/miniantags/abduction/machinery/console.dm]: well, this should only have the layer once, haha.

File: code/modules/mob/inventory.dm [hunk 5]
```diff
@@ -1,178 +1,291 @@
...
-/mob/proc/drop_l_hand()
-	if(!loc || !loc.allow_drop())
-		return
-	return unEquip(l_hand) //All needed checks are in unEquip
+	return FALSE
 
 
-//Drops the item in our right hand
+/*
+/mob/proc/drop_l_hand() //pending re-implementation
```
Comment (RemieRichards) [inline code/modules/mob/inventory.dm]: probably not actually, I don't think there's much use for this proc anymore (99% of uses were both procs at once), and I don't think there's use for a proc that drops all left items, or all right items.
tl;dr old comment.

File: code/game/gamemodes/clock_cult/clock_unsorted.dm (hunk 3 above)
Comment (ChangelingRain) [inline code/game/gamemodes/clock_cult/clock_unsorted.dm]: rip me

File: code/modules/mob/living/carbon/alien/humanoid/update_icons.dm [hunk 6]
```diff
@@ -68,27 +68,26 @@
 		apply_overlay(HANDCUFF_LAYER)
 
 //Royals have bigger sprites, so inhand things must be handled differently.
-/mob/living/carbon/alien/humanoid/royal/update_inv_r_hand()
+/mob/living/carbon/alien/humanoid/royal/update_inv_hands()
```
Comment (RemieRichards) [inline code/modules/mob/living/carbon/alien/humanoid/update_icons.dm]: this proc is now pretty much identical to the ```/carbon``` one, so I can remove them, that also means aliens will support multi-hands too.

File: code/modules/mob/living/carbon/carbon.dm [hunk 7]
```diff
@@ -79,38 +79,41 @@
 		return shock_damage
 
 
-/mob/living/carbon/swap_hand()
-	var/obj/item/item_in_hand = src.get_active_hand()
+/mob/living/carbon/swap_hand(held_index)
+	if(!held_index)
+		held_index = (held_index % held_items.len)+1
```
Comment (RemieRichards) [inline code/modules/mob/living/carbon/carbon.dm]: I do need to look at this bit, I was overzealous so this line probably exists in too many places, it also sometimes fails to go up an index, so it might just be wrong too.

**Comment by Kromgar** [comment]
http://vignette2.wikia.nocookie.net/vsbattles/images/b/b7/Six_armed_vajra_asura_by_superkaijuking-d8xsp1i.png/revision/latest?cb=20160314224350

File: code/modules/mob/living/carbon/inventory.dm [hunk 8]
```diff
@@ -1,4 +1,4 @@
-/mob/living/carbon/get_item_by_slot(slot_id)
+/mob/living/carbon/get_item_by_slot(slot_id, hand_index)
```
Comment (RemieRichards) [inline code/modules/mob/living/carbon/inventory.dm]: this arg isn't necessary, holdover from earlier version.

**Comment by Armhulen** [comment]
finally hivebot gets the love it deserves

File: code/modules/mob/living/say.dm [hunk 9]
```diff
@@ -266,14 +266,15 @@ var/list/crit_allowed_modes = list(MODE_WHISPER,MODE_CHANGELING,MODE_ALIEN)
 /mob/living/proc/radio(message, message_mode, list/spans)
 	switch(message_mode)
 		if(MODE_R_HAND)
-			if (r_hand)
-				r_hand.talk_into(src, message, , spans)
-			return ITALICS | REDUCE_RANGE
-
+			for(var/obj/item/r_hand in get_held_items_for_side("r", all = TRUE))
```
Comment (RemieRichards) [inline code/modules/mob/living/say.dm]: this means you speak into ALL radios in your right hands, might be better to just pick the first.

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 10]
```diff
@@ -24,6 +24,7 @@
 	var/obj/screen/inv3 = null
 	var/obj/screen/lamp_button = null
 	var/obj/screen/thruster_button = null
+	var/obj/screen/hands = null
```
Comment (RemieRichards) [inline code/modules/mob/living/silicon/robot/robot.dm]: this was on all mobs, it was only used on borgs, it's now only on borgs.

**Comment by xxalpha** [comment]
Madness

**Comment by GunHog** [comment]
I love it when RemRem brings the fun!

**Comment by PKPenguin321** [comment]
Oh hell yeah

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 11]
```diff
@@ -72,25 +72,19 @@
...
 
-	if(l_hand && !istype(l_hand, /obj/item/clothing))
-		var/final_block_chance = l_hand.block_chance - (Clamp((armour_penetration-l_hand.armour_penetration)/2,0,100)) + block_chance_modifier //So armour piercing blades can still be parried by other blades, for example
-		if(l_hand.hit_reaction(src, attack_text, final_block_chance, damage, attack_type))
-			return 1
-	if(r_hand && !istype(r_hand, /obj/item/clothing))
-		var/final_block_chance = r_hand.block_chance - (Clamp((armour_penetration-r_hand.armour_penetration)/2,0,100)) + block_chance_modifier //Need to reset the var so it doesn't carry over modifications between attempts
-		if(r_hand.hit_reaction(src, attack_text, final_block_chance, damage, attack_type))
-			return 1
+	for(var/obj/item/I in held_items)
```
Comment (KorPhaeron) [inline code/modules/mob/living/carbon/human/human_defense.dm]: Trying to think of a non terrible way to stop people from just carrying 3 riot shields and a gun

Or a double esword, shield, and gun

Comment (Alek2ander) [inline code/modules/mob/living/carbon/human/human_defense.dm]: Dismemberment

**Comment by AnturK** [comment]
Neat. One of these things that looked good on vg

**Comment by MrStonedOne** [comment]
get_held_index_of_item is having some issues, causing it to use a fuck ton of cpu when something burns in lava that was dropped from a stun (from lava)

It's locking up the mc, causing special issues, might be a stack overflow runtime

**Comment by Ohlos** [comment]
Wew, wow, woah. AWESOME.

**Comment by AdamElTablawy** [comment]
wizard with 4 spellblades named general grevious when

**Comment by KorPhaeron** [comment]
@AdamElTablawy 

Click delays remove most of the practical benefit of quad(or more) wielding other thank stacking shields though.

**Comment by AdamElTablawy** [comment]
but on the other hand FOUR ARMED WIZARD WITH SPELLBLADES NAMED GENERAL GREVIOUS AND A LICH.

**Comment by Supermichael777** [comment]
double e-sword and two riot-shields doesn't sound cancerous at all

**Comment by Shadowlight213** [comment]
Gilgamesh swordmaster antag when?

File: code/modules/mob/living/say.dm (hunk 9 above)
Comment (Shadowlight213) [inline code/modules/mob/living/say.dm]: Does this even work? I've never gotten the :r or :l commands to work.

**Comment by RemieRichards** [comment]
@MrStonedOne I've looked into it, there really isn't anything obvious on Lava's end or on ```unEquip()``` (the most logical call of  ```get_held_index_of_item()``` for something dying to lava), and I've checked every other call and there's no reason for it to be called once-per second in those.

**Comment by MrStonedOne** [comment]
it's not once per second, its taking 1 second per call based on the profile. the reason for that ended up being the stack overflow, stack overflow errors are different from runtime errors, they cause the entire call chain up to that point to end, going all the way back to the MC that started the call. "crashing" the stack causes the entire stack's time to count as only one call on the final proc of the stack (including the mc's time since it called the proc) in the profile. so basically get_hend_index_of_item was getting the self cpu of all those equip calls added to it's own self cpu. ie, it was a red harring.

the issue is something in unequip is causing unequip to get called again, seemingly thru shread or fire_act.

**Comment by RemieRichards** [comment]
perhaps it's ```forceMove()```? it used to be ```I.loc = loc``` but I changed it to fit standards ```I.forceMove(loc)```, but I suppose somewhere in the bowels of the game item's ```forceMove()``` or it's ```exited()``` is also calling ```unEquip()```

**Comment by MrStonedOne** [comment]
Oh god damn it! @Iamgoofball was right, lava code is fucked.

lava's turf entered calls burn_stuff, that loops thru the things on the turf (this is where it's fucked, it loops on 1 item entering for each item entering, rather then just act on that item), when it gets to a mob, it would call adjust fire stacks on the mob, that calls update_health(), that calls shred_clothing, that calls backpack.shred() that calls unequip, that calls turf entered(), rinse, repeat.

Turf's lava's entered needs to only act on the thing being entered, an process can act on everything in a loop. I'll fix that in a pr.

**Comment by RemieRichards** [comment]
@MrStonedOne s-so it's not my fault? yay! 
I've almost finished the lizard sprite fix, and I'm investigating the whole "stacks disappearing" (which I can't reproduce atm) so with the lizard+lava fixes done this might not be the death of the PR afterall.

**Comment by RemieRichards** [comment]
Anyone know a way to mass rename parts of a .dmi? ;_;
changing all the 25/23/2 into BEHIND/ADJ/FRONT for all mutant icon states is pain.

**Comment by RemieRichards** [comment]
I fixed lizards/species accessories.
I'm sorry crew.

xenoscum

**Comment by RemieRichards** [comment]
Oh, I've reproduced the stack deletion bug, but it might not be my fault, a stack can apparently ```merge()``` with itself, which for high enough amounts of a stack is fine, it'll use itself and then refill itself completely, if the value is too low however you'll just end up consuming the stack into nothingness.

**Comment by FantasticFwoosh** [comment]
I smell a sundowner double handheld e-sword and two energy shields meme

**Comment by Iamgoofball** [comment]
IM ALWAYS RIGHT

**Comment by Razharas** [comment]
I wish all our inventory was buildable from bodyparts

**Comment by Bluespace** [comment]
_General Grievous:_ Jedi. You are surrounded. Your army is decimated. Make peace with the force, now. For this is your final hour. But know that I, General Grievous, am not completely without mercy. I will grant you a warrior's death. Prepare!

**Comment by Alek2ander** [comment]
botw23

**Comment by as334** [comment]
How does this work with health dolls?

**Comment by RemieRichards** [comment]
@as334 it'll display all (left/right seperate) arms in the same spot, until they're all dismembered, then it'll show the usual missing limb.

**Comment by RemieRichards** [comment]
@tgstation/commit-access anytime guys.

**Comment by KorPhaeron** [comment]
Has it been reviewed?

**Comment by RemieRichards** [comment]
@KorPhaeron No, what do you think the ping is for, duuuuh.

**Comment by KorPhaeron** [comment]
Coulda just been merge begging

File: code/modules/mob/living/carbon/human/human_defines.dm [hunk 12]
```diff
@@ -53,3 +53,5 @@ var/global/default_martial_art = new/datum/martial_art
 
 	var/drunkenness = 0 //Overall drunkenness - check handle_alcohol() in life.dm for effects
 	var/datum/personal_crafting/handcrafting
+
+	var/list/hand_organs = list() //a collection of arms (or actually whatever the fug /bodyparts you monsters use to wreck my systems)
```
Comment (phil235) [inline code/modules/mob/living/carbon/human/human_defines.dm]: change the name to hand_bodyparts.

File: code/modules/mob/inventory.dm [hunk 13]
```diff
@@ -274,4 +369,47 @@
...
+			style = ui_style2icon(client.prefs.UI_style)
+		hud_used.build_hand_slots(style)
+
+
+/mob/living/carbon/human/change_number_of_hands(amt)
+	var/old_limbs = held_items.len
+	if(amt < old_limbs)
+		for(var/i in hand_organs.len to amt step -1)
+			var/obj/item/bodypart/BP = hand_organs[i]
+			BP.dismember()
```
Comment (phil235) [inline code/modules/mob/inventory.dm]: dismember() means a messy dismembering, if you just want to lose a limb then you have drop_limb()

**Comment by phil235** [comment]
This is going to give lots of conflict to my dismemberment PR :cry:

File: code/modules/mob/inventory.dm (hunk 13 above)
Comment (RemieRichards) [inline code/modules/mob/inventory.dm]: No, I want it messy atm.

File: code/modules/mob/living/carbon/human/human_defines.dm (hunk 12 above)
Comment (RemieRichards) [inline code/modules/mob/living/carbon/human/human_defines.dm]: sure, probably not tonight tho, stressed.

**Comment by RemieRichards** [comment]
Shouldn't be major @phil235, I don't actually touch bodyparts much.

**Comment by RemieRichards** [comment]
@Cheridan / @tgstation/commit-access 
pls guys.

**Comment by nullbear** [comment]
>  **RemieRichards:** I don't actually touch bodyparts much.

Doesn't that get a tad bit lonely sometimes?

**Comment by RemieRichards** [comment]
@nullbear ;_:
